File: code/R/R Computational Toolbox Files/SocialNetworks R/SocNet.R
# Social Networks

################################################################################################################
# read file
# Household ID	Person ID	Activity ID	Purpose	Start Time	Duration	Location
activities = as.matrix(read.table('sample_activities.dat'));
###activities = as.matrix(read.table('activities-portland-1-v1.dat'));

# generate list of nodes for people, identified by integer

maxPersonId = activities[nrow(activities),2]
personIdLst = 1:maxPersonId;

# get list of locations

###locationIDLst = genLocIDLst(activities);
locationIDLst = t(unique(activities[1:nrow(activities),7]));

# function to return index of location in locationIDLst
locationIndex<-function( loc, locationIDLst ) {
   return(which(locationIDLst == loc))
}

locationIndex(8159, locationIDLst)

# generate adjacency matrix with people indices as row labels and 
# location indices as column labels
# genPeopleLocConnMat - function to generate people-to-location graph
# adjMat = genPeopleLocConnMat( maxPersonId, locationIDLst, activities )
connMat = genPeopleLocConnMat( maxPersonId, locationIDLst, activities )

################################################################################################################

## MINIMUM DOMINATING SET PROBLEM

# function to return the degree of a person node

degPerson<-function(i) {
	return (sum(adjMat[i, 1:ncol(adjMat)]))
	}

# function to return the degree of a location node
degLocation<-function(j) {
	#return (sum(adjMat[1:nrow(adjMat), j])) ### this line had adjMat in the previous one and now it is connMat, is that a change?
	return (sum(connMat[1:nrow(connMat), j]))
	}

# list of ordered pairs of location index & corresponding degree
lst = 1:length(locationIDLst)
locDegPairLst = matrix(0,nrow=length(locationIDLst),ncol=2)

for (i in 1:length(locationIDLst)) {
    locDegPairLst[i,1] = i
    locDegPairLst[i,2] = degLocation(i)
}
print(locDegPairLst)

## function to return lst sorted by the second members of the ordered pairs
sortSecond <-function(locaDegPairLst){return(locDegPairLst[order(locDegPairLst[,2]),])
}


## Function to return list of personIDs adjacent to location loc

adjacentPeopleLst<-function(loc){
	 #return(t(which(adjMat[,loc]==1)))
	 return(t(which(connMat[,loc]==1)))
		}
adjacentPeopleLst(3)
adjacentPeopleLst(4)
 for (i in 1:52){
     adjacentPeopleLst(i)}

# test
locations = minDominating(locationIDLst,connMat,1)
length(locations)
print(locations)
# test 
locations = minDominating(locationIDLst,connMat,0.5)
length(locations)
print(locations)
# test
locations = minDominating(locationIDLst,connMat,0.75)
length(locations)
print(locations)



##################################################################################
#### People-to-people graph and degree distribution of people-to-people graph

#adjPeopleMat = personToPerson(adjMat)
#adjPeopleMat = personToPerson(adjMat)
connPeopleMat = peopleToPeople(connMat)

##use function to return the degree of a person node in people-to-people graph
#distribLst = degPersonPPG(adjPeopleMat, 1:maxPersonId)
distribLst = c()
for (i in 1:maxPersonId){
	distribLst = append(distribLst,degPersonPPG(connPeopleMat,i))
			}

#print(distribLst)
hist(distribLst,col="grey")

# # average degree
 mean(distribLst)    

# #################################################################################
# ######## Clustering coefficient

## Function to return list of personIDs adjacent to person v in person-to-person 
## graph
# adjacentPeople = @(adjPeopleMat, v) find(adjPeopleMat(v,:));-->>ask about this one as well???


adjacentPeople <-function(connPeopleMat, v) { #adjPeopleMat[v,]

return(which(connPeopleMat[v,]==1))
}

                      print(adjacentPeople(connPeopleMat,1))
for (i in 1:maxPersonId){
     adjacentPeople(connPeopleMat, i)
 }

 adjacentPeople(connPeopleMat, 18)


## Test of function to return the number of edges  in a set in person-to-person graph
# numPeopleEdges( adjPeopleMat, [1 2 3])

numPeopleEdges(connPeopleMat, c(1, 2, 3))

clusteringCoeff(connPeopleMat, 18)

 coeffLst = c();
 for (i in 1:18){
     #coeffLst = [coeffLst, clusteringCoeff(adjPeopleMat, i)];
     coeffLst = 0#cbind(coeffLst, clusteringCoeff(adjPeopleMat, i));
    }
coeffLst
mean(coeffLst)

File: products/Introduction_to_Computational_Science_Modules/02_System_Dynamics/MATLAB/Files/SocialNetworksMATLAB/SocNet.m
% Social Networks

clear all
close all
clc

%% Connection Matrix
% read file
% Household ID	Person ID	Activity ID	Purpose	Start Time	Duration	Location
activities = load('sample_activities.dat');
%%%activities = [4 7 1 0 0 13950 2938; 4 7 2 1 27000 17550 27618; 4 7 3 4 64800 29250 2938; 4 8 4 0 0 17400 2938; 4 8 5 2 34200 3060 6270; 4 8 6 2 38400 3060 21032; 4 8 7 0 43200 3150 2938; 4 8 8 2 47099 4215 15370; 4 8 9 4 54299 34500 2938; 5 9 1 0 0 13350 10628; 5 9 2 1 25800 19049 29740; 5 9 3 4 61800 30749 10628; 6 18 25 0 0 15300 2938; 6 18 26 7 30600 18149 5212; 6 18 27 4 65700 1800 2938; 6 18 28 2 67860 2388 19815; 6 18 29 0 70499 26400 2938];
%%%activities = load('activities-portland-1-v1.dat');

% generate list of nodes for people, identified by integer
maxPersonId = activities(end, 2)
personIdLst = 1:maxPersonId;

% get list of locations
genLocIDLst = @(activities)  unique(activities(:, 7)');

locationIDLst = genLocIDLst(activities);

% function to return index of location in locationIDLst
index(8159, locationIDLst)

% generate adjacency matrix with people indices as row labels and 
% location indices as column labels
% genPeopleLocConnMat - function to generate people-to-location graph
connMat = genPeopleLocConnMat( maxPersonId, locationIDLst, activities );

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Minimum dominating set problem

% function to return the degree of a location node
degLocation = @(j) sum(connMat(:, j));

% list of ordered pairs of location index & corresponding degree
lst = 1:length(locationIDLst);
locDegPairLst = [lst; degLocation(lst)]'

% function to return lst sorted by the second members of the ordered pairs
sortSecond = @(lst) sortrows(locDegPairLst, -2)

sortSecond(locDegPairLst)

% Function to return list of personIDs adjacent to location loc
adjacentPeopleLst = @(loc) find(connMat(:, loc) == 1)';

adjacentPeopleLst(3)
adjacentPeopleLst(4)
for i=1:52
    adjacentPeopleLst(i)
end

% test
locations = minDominating(locationIDLst, connMat, 1)
length(locations)

% test
locations = minDominating(locationIDLst, connMat, 0.5)
length(locations)

% test
locations = minDominating(locationIDLst, connMat, 0.75)
length(locations)

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% People-to-people graph and  degree distribution of people-to-people graph
connPeopleMat = peopleToPeople(connMat);

% use function to return the degree of a person node in people-to-people graph
distribLst = degPersonPPG(connPeopleMat, 1:maxPersonId)

hist(distribLst)

% average degree
mean(distribLst)

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Clustering coefficient

% Function to return list of personIDs adjacent to person v in person-to-person 
% graph
adjacentPeople = @(connPeopleMat, v) find(connPeopleMat(v,:));

% test function
for i = 1:maxPersonId
    adjacentPeople(connPeopleMat, i)
end

adjacentPeople(connPeopleMat, 18)

% Test of function to return the number of edges  in a set in person-to-person graph
numPeopleEdges( connPeopleMat, [1 2 3])

clusteringCoeff(connPeopleMat, 18)

coeffLst = [];
for i = 1:18
    coeffLst = [coeffLst, clusteringCoeff(connPeopleMat, i)];
end;
coeffLst
mean(coeffLst)